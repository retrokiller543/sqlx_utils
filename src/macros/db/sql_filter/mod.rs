mod sql_delimiter;
mod sql_impl;
mod sql_operator;
