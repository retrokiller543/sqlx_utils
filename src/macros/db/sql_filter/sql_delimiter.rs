#[macro_export]
macro_rules! sql_delimiter {
    {
        $(#[$struct_meta:meta])*
        $vis:vis struct $ident:ident $(<
            $($lt:lifetime,)*
            $($generic:ident $(: $($generic_bound:tt)+)?),*
            $(,)?
        >)? {
            $(
                $(#[$field_meta:meta])*
                $field_vis:vis $field_ident:ident: $field_ty:ty
            ),* $(,)?
        }

        $apply_filter:ident ($apply_self:ident, $builder:ident) $apply_block:block
        $should_apply_filter:ident($should_self:ident) $should_apply_block:block
        $(where $($where_clause:tt)+)?
    } => {
        $(#[$struct_meta])*
        $vis struct $ident$(<$($lt,)* $($generic),*>)? {
            $(
                $(#[$field_meta])*
                $field_vis $field_ident: $field_ty
            ),*
        }

        impl$(<$($lt,)* $($generic),*>)? $ident$(<$($lt,)* $($generic),*>)?
        $(where $($where_clause)+)?
        {
            #[inline]
            $vis fn new($($field_ident: $field_ty),*) -> Self {
                Self { $($field_ident),* }
            }
        }

        $crate::sql_impl! {
            $ident$(<$($lt,)* $($generic),*>)?;
            $apply_filter($apply_self, $builder) $apply_block
            $should_apply_filter($should_self) $should_apply_block
            $(where
                $($generic: $crate::traits::SqlFilter<'args>),*)?
        }
    };
}