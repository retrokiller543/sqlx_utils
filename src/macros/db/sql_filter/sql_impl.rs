#[macro_export]
macro_rules! sql_impl {
    // Base case with explicit lifetime bounds and where clauses
    {
        $ident:ident $(<
            $($lt:lifetime,)*
            $($generic:ident $(: $($generic_bound:tt)+)?),*
            $(,)?
        >)?;
        $apply_filter:ident ($apply_self:ident, $builder:ident) $apply_block:block
        $should_apply_filter:ident($should_self:ident) $should_apply_block:block
        $(where $($where_clause:tt)+)?
    } => {
        impl<'args $(, $($lt,)* $($generic),*)?> $crate::traits::SqlFilter<'args>
            for $ident$(<$($lt,)* $($generic),*>)?
        where
            $($($generic: $($($generic_bound)+ +)? 'args,)*)?
            $($($where_clause)+)?
        {
            #[inline]
            fn $apply_filter(self, builder: &mut ::sqlx::QueryBuilder<'args, $crate::types::Database>) {
                let filter_impl = |$apply_self: Self, $builder: &mut ::sqlx::QueryBuilder<'args, $crate::types::Database>|
                    $apply_block;
                filter_impl(self, builder)
            }

            #[inline]
            fn $should_apply_filter(&self) -> bool {
                let should_apply_impl = |$should_self: &Self| $should_apply_block;
                should_apply_impl(self)
            }
        }
    };

    // Shorthand case for simple SqlFilter bounds
    {
        $ident:ident $(<$($lt:lifetime,)* $($generic:ident),* $(,)?>)?;
        $apply_filter:ident ($apply_self:ident, $builder:ident) $apply_block:block
        $should_apply_filter:ident($should_self:ident) $should_apply_block:block
    } => {
        $crate::traits::sql_impl! {
            $ident$(<$($lt,)* $($generic: $crate::traits::SqlFilter<'args>),*>)?;
            $apply_filter($apply_self, $builder) $apply_block
            $should_apply_filter($should_self) $should_apply_block
            where
        }
    };

    {
        $ident:ident $(<$($lt:lifetime)? $(,)? $($generic:ident),* $(,)?>)?;
        $apply_filter:ident (_, _) $apply_block:block
        $should_apply_filter:ident(_) $should_apply_block:block
    } => {
        impl<'args, $($($lt,)? $($generic),*)?> $crate::traits::SqlFilter<'args> for $ident$(<$($lt,)? $($generic),*>)?
        $(where
            $($generic: $crate::traits::SqlFilter<'args> + 'args),*)?
        {
            #[inline]
            fn $apply_filter(self, builder: &mut ::sqlx::QueryBuilder<'args, $crate::types::Database>) {
                let filter_impl = |_: Self, _: &mut ::sqlx::QueryBuilder<'args, $crate::types::Database>| $apply_block;
                filter_impl(self, builder);
            }

            #[inline]
            fn $should_apply_filter(&self) -> bool {
                let should_apply_impl = |_: &Self| $should_apply_block;
                should_apply_impl(self)
            }
        }
    };
}
