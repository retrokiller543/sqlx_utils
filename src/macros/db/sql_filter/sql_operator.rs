#[doc(hidden)]
#[macro_export]
macro_rules! sql_operator {
    ($vis:vis $ident:ident, $lit:literal) => {
        $vis struct $ident<T> {
            column: &'static str,
            value: Option<T>,
        }

        impl<'args, T> $ident<T>
        where
            T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database> + 'args,
        {
            #[inline]
            $vis fn new(column: &'static str, value: Option<T>) -> Self {
                Self { column, value }
            }
        }

        impl $ident<$crate::filter::Raw> {
            #[inline]
            $vis fn new_raw(column: &'static str, value: $crate::filter::Raw) -> Self {
                Self { column, value: Some(value) }
            }
        }

        $crate::sql_impl! {
            $ident<T>;

            apply_filter(s, builder) {
                if let Some(val) = s.value {
                    builder.push(s.column);
                    builder.push(concat!(" ", $lit, " "));
                    builder.push_bind(val);
                }
            }

            should_apply_filter(s) {
                s.value.is_some()
            }

            where
                T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database>
        }

        impl<'args> $crate::traits::SqlFilter<'args> for $ident<$crate::filter::Raw>
        {
            #[inline]
            fn apply_filter(self, builder: &mut ::sqlx::QueryBuilder<'args, $crate::types::Database>) {
                if let Some(val) = self.value {
                    builder.push(self.column);
                    builder.push(concat!(" ", $lit, " "));
                    val.apply_filter(builder)
                }

            }

            #[inline]
            fn should_apply_filter(&self) -> bool {
                self.value.is_some()
            }
        }

        ::paste::paste! {
            #[inline]
            pub fn [< $ident:snake >]<'args, T>(
                column: &'static str,
                value: Option<T>
            ) -> $crate::filter::Filter<$ident<T>>
            where
                T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database> + 'args,
            {
                $crate::filter::Filter::new($ident::new(column, value))
            }

            #[inline]
            pub fn [< $ident:snake _raw >](
                column: &'static str,
                value: $crate::filter::Raw
            ) -> $crate::filter::Filter<$ident<$crate::filter::Raw>> {
                $crate::filter::Filter::new($ident::new_raw(column, value))
            }
        }
    };

    ($vis:vis $ident:ident<$ty:ty>, $lit:literal) => {
        $vis struct $ident {
            column: &'static str,
            value: Option<$ty>,
        }

        impl $ident {
            #[inline]
            $vis fn new(column: &'static str, value: Option<impl Into<$ty>>) -> Self {
                let value = value.map(::core::convert::Into::into);
                Self { column, value }
            }
        }

        $crate::sql_impl! {
            $ident;

            apply_filter(s, builder) {
                if let Some(value) = s.value {
                    builder.push(s.column);
                    builder.push(concat!(" ", $lit, " "));
                    builder.push_bind(value);
                }
            }

            should_apply_filter(s) {
                s.value.is_some()
            }
        }

        ::paste::paste! {
            #[inline]
            pub fn [< $ident:snake >](
                column: &'static str,
                value: Option<impl Into<$ty>>
            ) -> $crate::filter::Filter<$ident> {
                $crate::filter::Filter::new($ident::new(column, value))
            }

            #[inline]
            pub fn [< $ident:snake _raw >](
                column: &'static str,
                value: $crate::filter::Raw
            ) -> $crate::filter::Filter<$ident> {
                $crate::filter::Filter::new($ident::new(column, Some(value.0)))
            }
        }
    };

    ($vis:vis $ident:ident[], $lit:literal) => {
        $vis struct $ident<T> {
            column: &'static str,
            values: Vec<T>,
        }

        impl<'args, T> $ident<T>
        where
            T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database> + 'args,
        {
            #[inline]
            $vis fn new(column: &'static str, values: impl IntoIterator<Item = T>) -> Self {
                let values = values.into_iter().collect();
                Self { column, values }
            }
        }

        $crate::sql_impl! {
            $ident<T>;

            apply_filter(s, builder) {
                if !s.should_apply_filter()  {
                    return;
                }

                builder.push(s.column);
                builder.push(concat!(" ", $lit, " ("));

                let mut first = true;
                for val in s.values {
                    if !first {
                        builder.push(", ");
                    }
                    builder.push_bind(val);
                    first = false;
                }

                builder.push(")");
            }

            should_apply_filter(s) {
                !s.values.is_empty()
            }

            where
                T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database>
        }

        ::paste::paste! {
            #[inline]
            pub fn [< $ident:snake >]<'args, T>(
                column: &'static str,
                values: impl IntoIterator<Item = T>
            ) -> $crate::filter::Filter<$ident<T>>
            where
                T: ::sqlx::Type<$crate::types::Database> + ::sqlx::Encode<'args, $crate::types::Database> + 'args,
            {
                $crate::filter::Filter::new($ident::new(column, values))
            }
        }
    };

    ($vis:vis $ident:ident) => {
        $vis struct $ident;

        impl Default for $ident {
            fn default() -> Self {
                Self::new()
            }
        }

        impl $ident {
            #[inline]
            $vis fn new() -> Self {
                Self
            }
        }

        $crate::sql_impl! {
            $ident;
            apply_filter(_, _) {}
            should_apply_filter(_) { false }
        }

        ::paste::paste! {
            #[inline]
            pub fn [< $ident:snake >]() -> $crate::filter::Filter<$ident> {
                $crate::filter::Filter::new($ident::new())
            }

            #[inline]
            pub fn [< $ident:snake _raw >]() -> $crate::filter::Filter<$ident> {
                $crate::filter::Filter::new($ident::new())
            }
        }
    };
}
