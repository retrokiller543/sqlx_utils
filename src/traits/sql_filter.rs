//! Sql filtering trait for defining type safe dynamic filters.

use crate::types::Database;
use sqlx::{Database as DatabaseTrait, QueryBuilder};

/// Trait for creating SQL filter conditions that can be applied to database queries.
///
/// The `SqlFilter` trait provides a standardized way to define reusable, type-safe
/// SQL filter conditions. Filters can be composed and combined using logical operators
/// to create complex query criteria.
///
/// # Type Parameters
///
/// * `'args`: The lifetime for query arguments
/// * `DB`: The database backend type (defaults to the configured default database)
///
/// # Examples
///
/// Basic implementation using the macro:
/// ```rust
/// # use sqlx_utils::sql_filter;
/// # use sqlx_utils::traits::SqlFilter;
/// # use sqlx_utils::types::Database;
/// # use sqlx::QueryBuilder;
///
/// sql_filter! {
///     pub struct UserFilter {
///         SELECT * FROM users WHERE
///         id = i32 AND
///         ?name LIKE String AND
///         ?email LIKE String
///     }
/// }
///
/// // Usage
/// # fn example() {
/// let filter = UserFilter::new(42)
///     .name("Alice%")
///     .email("alice@example.com");
///
/// let mut builder = QueryBuilder::<Database>::new("SELECT * FROM users WHERE ");
/// filter.apply_filter(&mut builder);
/// # }
/// ```
///
/// Custom implementation:
/// ```rust
/// # use sqlx_utils::traits::SqlFilter;
/// # use sqlx_utils::types::Database;
/// # use sqlx::QueryBuilder;
///
/// struct AgeRangeFilter {
///     min_age: Option<i32>,
///     max_age: Option<i32>,
/// }
///
/// impl<'args> SqlFilter<'args, Database> for AgeRangeFilter {
///     fn apply_filter(self, builder: &mut QueryBuilder<'args, Database>) {
///         if self.min_age.is_some() || self.max_age.is_some() {
///             let mut first = true;
///
///             if let Some(min) = self.min_age {
///                 if !first { builder.push(" AND "); }
///                 builder.push("age >= ");
///                 builder.push_bind(min);
///                 first = false;
///             }
///
///             if let Some(max) = self.max_age {
///                 if !first { builder.push(" AND "); }
///                 builder.push("age <= ");
///                 builder.push_bind(max);
///             }
///         }
///     }
///
///     fn should_apply_filter(&self) -> bool {
///         self.min_age.is_some() || self.max_age.is_some()
///     }
/// }
/// ```
///
/// # Implementation Notes
///
/// When implementing this trait:
///
/// 1. The [`apply_filter`](SqlFilter::apply_filter) method should add SQL conditions to the builder
/// 2. The [`should_apply_filter`](SqlFilter::should_apply_filter) method should return `true` if this filter has criteria to apply
/// 3. Consider using the [`sql_filter!`](crate::sql_filter) macro for common filter patterns
/// 4. Ensure proper parameterization to prevent SQL injection
#[diagnostic::on_unimplemented(
    message = "The filter type `{Self}` must implement `SqlFilter<'args>` to be used in queries",
    label = "this type does not implement `SqlFilter<'args>`",
    note = "Type `{Self}` does not implement the `SqlFilter` trait with the required lifetime `'args`.",
    note = "SqlFilter is implemented by default for the `sqlx_utils::types::Database` database, things might not work if you are using a custom database."
)]
pub trait SqlFilter<'args, DB: DatabaseTrait = Database> {
    /// Applies this filter's conditions to a SQL query builder.
    ///
    /// This method should add the necessary SQL conditions represented by this filter
    /// to the provided query builder. It should handle binding parameters securely.
    ///
    /// # Parameters
    ///
    /// * `self` - The filter instance, consumed during application
    /// * `builder` - The query builder to which the filter will be applied
    fn apply_filter(self, builder: &mut QueryBuilder<'args, DB>);

    /// Determines whether this filter has meaningful conditions to apply.
    ///
    /// This method should return `true` if the filter has non-default conditions
    /// that should be included in a query, or `false` if the filter is empty or
    /// represents default criteria that don't need to be applied.
    ///
    /// # Returns
    ///
    /// * `true` - If the filter has conditions to apply
    /// * `false` - If the filter has no conditions to apply
    fn should_apply_filter(&self) -> bool;
}
