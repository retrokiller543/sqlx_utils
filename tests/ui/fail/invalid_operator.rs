use sqlx_utils::sql_filter;

sql_filter! {
    pub struct UserFilter {
        SELECT * FROM users WHERE
        age INVALID i32  // Invalid operator
    }
}

fn main() {
    let _filter = UserFilter::new(25);
}
