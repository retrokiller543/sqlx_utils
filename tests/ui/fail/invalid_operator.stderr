error: Invalid SQL operator `INVALID`

         = help: Supported operators are: =, !=, >, <, >=, <=, LIKE, ILIKE, IN, NOT IN

 --> tests/ui/fail/invalid_operator.rs:6:13
  |
6 |         age INVALID i32  // Invalid operator
  |             ^^^^^^^
